File: CSharpCodeBase/aiactions/ai_avoidance.lua
local Class = require 'utils.hump.class'
local Timer = require 'utils.hump.timer'
local AINode = require 'utils.ailib.ainode'

local AI_Avoidance = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
end})

function AI_Avoidance:Visit()
  self.target = self.entity.goalProsecution
  if self.target then
    if self.status == NODE_READY then
      self.status = NODE_RUNNING
      self:Moving()
    elseif self.status == NODE_RUNNING then
      if not self.target.interactable then
        self.status = NODE_FAILURE
        return
      end
      if not self.entity.mover:IsHaveGoal() then
        self.status = NODE_SUCCESS
      end
    end
  end
  return self.status
end

function AI_Avoidance:Moving()
  self.entity.mover:ResetSpeed()
  local tmp = self.entity:GetPosition() - self.target:GetPosition()
  self.entity.mover:SetInput(tmp , 0, true)
end

return AI_Avoidance

File: CSharpCodeBase/aiactions/ai_avoidance.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_avoidance :ainode{
 public void init (self, entity);
   AINode.init(self, entity);
 }})
 public void Visit(){
   this.target = this.entity.goalProsecution;
   if(this.target  ){ 
     if(this.status == NODE_READY  ){ 
       this.status = NODE_RUNNING;
       self:Moving();
     }else{if this.status == NODE_RUNNING  ){ 
       if(not this.target.interactable  ){ 
         this.status = NODE_FAILURE;
         return;
       }
       if(not this.entity.mover:IsHaveGoal()  ){ 
         this.status = NODE_SUCCESS;
       }
     }
   }
   return this.status;
 }
 public void Moving(){
   this.entity.mover:ResetSpeed();
   var tmp = this.entity:GetPosition() - this.target:GetPosition();
   this.entity.mover:SetInput(tmp , 0, true);
 }
 }}